### DISCUSSION TITLE
continuous load test generate an error and the memory grows significantly

### DISCUSSION THREAD

**Discussion by ShangChien** [rdkit/rdkit-js#143] [Q&A] [answered]
I batch test in a `SharedWorker.js` using the function that gets the molecular svg image，code in sharedworker as following：
``` js
// import rdkit into the worker, and export the renderMol function
importScripts("./RDKit_minimal.js")
//initialize the rdkit
initRDKitModule().then(res=>{
	self.rdkit = res	
  self.rdkit.prefer_coordgen(true);
}).catch(err=>{
	console.log(err)
})


let i = 0 //counter for the number of times the worker is connected
let out = null //the output(svg image) of the renderMol function
function renderMol(props){
  let mol = self.rdkit.get_mol(props.smiles ?? "");
  let qmol = self.rdkit.get_qmol(props.qsmiles ?? "");
  let mdetailsRaw = mol.get_substruct_matche(qmol);
  let mDetail =  JSON.parse(mdetailsRaw) : [];
  mDetail = JSON.stringify(mDetail);
  out = mol.get_svg_with_highlights(mDetail);
  mDetail=null;
  mdetailsRaw=null;
  qmol=null ;
  mol=null;
  return out
}

self.onconnect = function(e) {
  var port = e.ports[0];
  console.log("the ",i++,"th time connected")
  port.onmessage = function(e) {
    out = renderMol(JSON.parse(e.data))
    port.postMessage(out);
		out=null;
  }
  port.onmessageerror = function(e) {
    console.log("error", e);
  }
}
self.onerror = function(e) {
  console.log("error", e);
}
```
I always getting the following error after run about 600th times, It seems to be related to memory
![image](https://user-images.githubusercontent.com/48350636/179452191-ebf17a65-c38b-4bbc-b197-26fc0242ca21.png)

**current solution:**
every time connected, re-run the rdkit init code(just let it in `self.onconnect` function). 
```js
...
self.onconnect = function(e) {
  initRDKitModule().then(res=>{
      self.rdkit = res	
  self.rdkit.prefer_coordgen(true);
  }).catch(err=>{
     console.log(err)
  })
  var port = e.ports[0];
  console.log("the ",i++,"th time connected")
  port.onmessage = function(e) {
    out = renderMol(JSON.parse(e.data))
    port.postMessage(out);
		out=null;
  }
  port.onmessageerror = function(e) {
    console.log("error", e);
  }
}
...
```
It can get rid of this error, but a bit slower than before.
It's funny, but I don't know why. 

my env ：
OS ：windows10 (19044.1826), Chrome
cpu: 11th Gen Intel(R) Core(TM) i5-11400 @ 2.60GHz   2.59 GHz
ram: 16 Gb
rdkitjs: 2022.3.3 and 2022.3.4-1.0.4

**Comment by MichelML** [comment, marked as answer]
@ShangChien thank you for your question,

since the mol and qmol objects are c++ objects coming from the wasm module, you need to delete them explicitly and not only nullify them.

See example here:
https://github.com/rdkit/rdkit-js/blob/master/examples/react/src/components/MoleculeStructure/MoleculeStructure.js#L92

See the emscripten reference here:
https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management

If you still get the error after changing:

```js
  qmol=null ;
  mol=null;
```

for 

```js
  qmol.delete();
  qmol = null;
  mol.delete();
  mol = null;
```

Let us know. Also, in my production code I have a small wrapper to avoid a double statement of deleting and nullifying all the time, like so:

```js
// it works for qmol too!
export const deleteMol = (mol) => {
  if (mol) {
    mol.delete();
    mol = null;
  }
};
```

Hope it helps.

cc maybe @ptosco can comment further here

**Reply by ptosco** [reply]
@ShangChien
@MichelML is absolutely right - make sure to always call `delete()` on the underlying WASM object to avoid memory leaks.

**Reply by ShangChien** [reply]
@ptosco @MichelML  Thank you for all reply, follow your effective suggestion, RDKitjs work very well in worker!
